packages/contracts/scripts/one-off/add-unregistered-contracts.ts

```diff
@@ -0,0 +1,206 @@
+    signer = await ethers.getSigner(ledgerAddress);
+    console.log("using wallet", await signer.getAddress());
+  } else {
+    const deployerWallet = getDeployerWallet();
+
+    if (!deployerWallet) {
+      throw new Error("Deployer wallet not found");
+    }
+
+    signer = deployerWallet.connect(ethers.provider);
+  }
+
+  let gnosisSetup: {
+    ethAdapter: EthersAdapter;
+    safeApiKit: SafeApiKit;
+    protocolKit: Safe;
+  } | null = null;
+
+  // Gnosis sdk setup
+  if (config.useGnosisSafe) {
+    const ethAdapter = new EthersAdapter({
+      ethers,
+      signerOrProvider: signer,
+    });
+
+    const safeApiKit = new SafeApiKit({
+      txServiceUrl: config.transactionServiceUrl,
+      ethAdapter,
+    });
+
+    const protocolKit = await Safe.create({
+      ethAdapter,
+      safeAddress: config.safeAddress,
+    });
+
+    gnosisSetup = {
+      ethAdapter,
+      safeApiKit,
+      protocolKit,
+    };
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // ACTION BEGINS HERE
+  //////////////////////////////////////////////////////////////////////////////
+
+  // Query for unregistered contracts
+  const registryAddresses = [
+    "0x2ee7b9bb2e038be7323a119701a191c030a61ec6",
+    "0x652490c8bb6e7ec3fd798537d2f348d7904bbbc2",
+  ];
+
+  const client = getClient();
+  const res = await client
+    .query(GetUnregisteredContractsDocument, {
+      registryAddresses,
+    })
+    .toPromise();
+
+  if (!res.data?.contracts_metadata || !res.data.contracts_metadata.length) {
+    console.log("No unregistered contracts found");
+    return;
+  }
+
+  const coreRegistry = CoreRegistryV1__factory.connect(
+    config.coreRegistryAddress,
+    signer
+  );
+
+  const txData: MetaTransactionData[] = [];
+  for (const contract of res.data.contracts_metadata) {
+    const contractCoreVersion = contract.core_version
+      ? ethers.utils.formatBytes32String(contract.core_version)
+      : ethers.constants.HashZero; // "0x0000000000000000000000000000000000000000000000000000000000000000"
+
+    const contractType = contract.contract_type
+      ? ethers.utils.formatBytes32String(contract.contract_type)
+      : ethers.constants.HashZero;
+
+    if (config.useGnosisSafe) {
+      const data = coreRegistry.interface.encodeFunctionData(
+        "registerContract",
+        [contract.address, contractCoreVersion, contractType]
+      );
+      txData.push({
+        to: config.coreRegistryAddress,
+        value: "0x00",
+        data,
+      });
+    } else {
+      try {
+        const tx = await coreRegistry.registerContract(
+          contract.address,
+          contractCoreVersion,
+          contractType
+        );
+        await tx.wait();
+        console.log(`Successfully registered contract: ${contract.address}`);
+      } catch (error) {
+        console.error(
+          `Failed to register contract ${contract.address}:`,
+          error
+        );
+      }
```

Comment:
see note about how this needs to go through the EngineFactory
ref https://github.com/ArtBlocks/artblocks-contracts/blob/main/packages/contracts/INFRASTRUCTURE.md#mainnet

additionally, recommend using registerMultipleContracts to stay gas efficient, as well as avoiding nonce errors in gnosis safe (the `gnosisSetup.protocolKit.getNonce()` function returns next nonce to execute, not next nonce to queue)

Reply:
thanks! refactored this to call registerMultipleContracts on the engine factory 🙏